File: algorithms/098.Validate_Binary_Search_Tree/Validate_Binary_Search_Tree.cpp
#include <bits/stdc++.h>

using namespace std;

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    bool dfs(TreeNode *root, long long minval, long long maxval) {
        if (root == NULL)
            return true;
        return root->val > minval &&
            root->val < maxval &&
            dfs(root->left, minval, root->val) &&
            dfs(root->right, root->val, maxval);
    }

public:
    bool isValidBST(TreeNode *root) {
        return dfs(root, (long long)INT_MIN - 1, (long long)INT_MAX + 1);
    }
};

int main() {

    return 0;
}

File: algorithms/098.Validate_Binary_Search_Tree/Validate_Binary_Search_Tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: True if the binary tree is BST, or false
    """
    def isValidBST(self, root):
        self.lastVal = None
        self.isBST = True
        self.validate(root)
        return self.isBST

    def validate(self, root):
        if root is None:
            return
        self.validate(root.left)
        if self.lastVal is not None and self.lastVal >= root.val:
            self.isBST = False
            return
        self.lastVal = root.val
        self.validate(root.right)

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution2:
    # @param root, a tree node
    # @return a boolean
    def isValidBST(self, root):

        def dfs(root, minval, maxval):
            if root is None:
                return True
            return minval < root.val < maxval and dfs(root.left, minval, root.val) and dfs(root.right, root.val, maxval)

        return dfs(root, -1<<32, 1<<32)
